skxray/cdi.py

```diff
@@ -96,32 +99,321 @@ def gauss(dims, sigma):
+        gaussian width used in sw algorithm.
+        default is 0.5.
+    sw_threshold : float, optional
+        shreshold cut in sw algorithm.
+        default is 0.1.
+    sw_start : float, optional
+        at which point to start to do shrinkwrap.
+        defualt is 0.2
+    sw_end : float, optional
+        at which point to stop shrinkwrap.
+        defualt is 0.8
+    sw_step : float, optional
+        the frequency to perform sw algorithm.
+        defualt is 10
+    n_iterations : int, optional
+        number of iterations to run.
+        default is 1000.
+
+    Returns
+    -------
+    obj_avg : array
+        reconstructed sample object
+    error_dict : dict
+        Error information for all iterations. The dict keys include
+        obj_error, diff_error and sup_error. Obj_error is a list of
+        the relative error of sample object. Diff_error is calculated as
+        the difference between new diffraction pattern and the original
+        diffraction pattern. And sup_error stores the size of the
+        sample support.
+    """
+
+    diffracted_pattern = np.array(diffracted_pattern)     # diffraction data
+
+    gamma_1 = -1/beta
+    gamma_2 = 1/beta
+
+    # get support index
+    sup_out_index = sup < 1
+
+    error_dict = {}
+    obj_error = np.zeros(n_iterations)
+    diff_error = np.zeros(n_iterations)
+    sup_error = np.zeros(n_iterations)
+
+    sup_old = np.zeros_like(diffracted_pattern)
+    obj_avg = np.zeros_like(diffracted_pattern).astype(complex)
+    avg_i = 0
+
+    time_start = time.time()
+    for n in range(n_iterations):
+        obj_old = np.array(sample_obj)
+
+        obj_a = pi_modulus(sample_obj, diffracted_pattern)
+        if pi_modulus_flag.lower() == 'real':
+            obj_a = np.abs(obj_a)
+
+        obj_a = (1 + gamma_2) * obj_a - gamma_2 * sample_obj
+        obj_a = pi_support(obj_a, sup_out_index)
+
+        obj_b = pi_support(sample_obj, sup_out_index)
+        obj_b = (1 + gamma_1) * obj_b - gamma_1 * sample_obj
+
+        obj_b = pi_modulus(obj_b, diffracted_pattern)
+        if pi_modulus_flag.lower() == 'real':
+            obj_b = np.abs(obj_b)
+
+        sample_obj += beta * (obj_a - obj_b)
+
+        # calculate errors
+        obj_error[n] = cal_relative_error(obj_old, sample_obj)
+        diff_error[n] = cal_diff_error(sample_obj, diffracted_pattern)
+
+        if sw_flag:
+            if((n >= (sw_start * n_iterations)) and (n <= (sw_end * n_iterations))):
+                if np.mod(n, sw_step) == 0:
+                    logger.info('Refine support with shrinkwrap')
+                    sup = find_support(sample_obj, sw_sigma, sw_threshold)
+                    sup_out_index = sup < 1
+                    sup_error[n] = np.sum(sup_old)
+                    sup_old = np.array(sup)
+
+        if n > start_avg*n_iterations:
+            obj_avg += sample_obj
+            avg_i += 1
+
+        logger.info('{} object_chi= {}, diff_chi={}'.format(n, obj_error[n],
```

Comment:
use old-style formatting here as it lets the logger be lazy and only do the formatting if strictly needed.  Something like

```python
logger.info("%s my message %s", 'foo', 'bar')
```

Note that the formatting inputs are args passed to the logging function.